libexec/bats-core/bats-gather-tests

```diff
@@ -0,0 +1,306 @@
+    BATS_PREVIOUS_RUNLOG_FILE="$BATS_RUN_LOGS_DIRECTORY/$BATS_PREVIOUS_RUNLOG_FILE"
+    if [[ $BATS_PREVIOUS_RUNLOG_FILE == "$BATS_RUNLOG_FILE" ]]; then
+      count=$(find "$BATS_RUN_LOGS_DIRECTORY" -name "$BATS_RUNLOG_DATE*" | wc -l)
+      BATS_RUNLOG_FILE="$BATS_RUN_LOGS_DIRECTORY/${BATS_RUNLOG_DATE}-$count.log"
+    fi
+    failed_tests=()
+    passed_tests=()
+    # store tests that were already filtered out in the last run for the same filter reason
+    last_filtered_tests=()
+    i=0
+    while read -rd $'\n' line; do
+      ((++i))
+      case "$line" in
+      "passed "*)
+        passed_tests+=("${line#passed }")
+        ;;
+      "failed "*)
+        failed_tests+=("${line#failed }")
+        ;;
+      "status-filtered $filter_status"*) # pick up tests that were filtered in the last round for the same status
+        last_filtered_tests+=("${line#status-filtered "$filter_status" }")
+        ;;
+      "status-filtered "*) # ignore other status-filtered lines
+        ;;
+      "#"*) # allow for comments
+        ;;
+      *)
+        printf "Error: %s:%d: Invalid format: %s\n" "$BATS_PREVIOUS_RUNLOG_FILE" "$i" "$line" >&2
+        exit 1
+        ;;
+      esac
+    done < <(sort "$BATS_PREVIOUS_RUNLOG_FILE")
+
+    # enable filter only, when there is something to filter
+    function should_skip_because_of_status() {
+      #echo "Match:" 
+      #echo "$line"
+      #printf "%s\n" "${failed_tests[@]}"
+      #echo "or"
+      #printf "%s\n" "${last_filtered_tests[@]}"
+      if bats_filter_test_by_status "$line"; then 
+        if ! bats_binary_search "$line" last_filtered_tests; then
+          included_tests+=("$line")
+          #echo "included"
+          return 1
+        fi
+      fi
+      #echo "excluded"
+      excluded_tests+=("$line")
+      return 0
+    }  >&2
+  else
+    printf "No recording of previous runs found. Running all tests!\n" >&2
+  fi
+fi
+
+# shellcheck source=lib/bats-core/tracing.bash
+source "$BATS_ROOT/lib/bats-core/tracing.bash"
+
+
+bats_gather_tests_exit_trap() {
+  local bats_gather_tests_exit_status=$?
+  trap - ERR EXIT DEBUG
+  if [[ ${BATS_ERROR_STATUS:-0} != 0 ]]; then
+    bats_gather_tests_exit_status=$BATS_ERROR_STATUS
+    printf "1..1\nnot ok 1 bats-gather-tests\n"
+    bats_get_failure_stack_trace stack_trace
+    bats_print_stack_trace "${stack_trace[@]}"
+    bats_print_failed_command "${stack_trace[@]}" 
+  fi >&2
+  exit "$bats_gather_tests_exit_status"
+}
+trap bats_gather_tests_exit_trap EXIT
+
+bats_set_stacktrace_limit
+
+bats_setup_tracing
+focus_mode=0
+for filename in "$@"; do
+  if [[ ! -f "$filename" ]]; then
+    abort 'Test file "%s" does not exist.\n' "${filename}"
+  fi
+
+  test_names=()
+  test_dupes=()
+
+  BATS_TEST_FILENAME="$filename"
+  _bats_test_functions_setup -1 # invalid TEST_NUMBER, as this is not a test
+
+  BATS_TEST_NAME=source
+  BATS_TEST_FILTER="$BATS_TEST_FILTER" bats_preprocess_source # uses BATS_TEST_FILENAME
```

Comment:
There doesn't seem to be anything left that's using `BATS_TEST_FILTER`.

Reply:
I tried to remove `BATS_TEST_FILTER` in https://github.com/bats-core/bats-core/pull/907